### PR TITLE
lab/assembly-intro: Add CF vs OF example: Fixes #123

### LINKED ISSUE #123: lab/assembly-intro: Add `CF` vs `OF` example
The "Intro to Assembly" lab only briefly mentions the carry and overflow flags [here](https://github.com/systems-cs-pub-ro/iocla/blob/56d3dccb47fea0f801281462b54e65aa0a944b08/laborator/content/introducere-asamblare/README.md?plain=1#L103-L110). We should add an exercise that showcases these 2. It doesn't need to be complicated. 4 simple 8-bit `add`s should suffice: one sets no flags, one for `CF`, one for `OF` and another one for both of them.

### PR DESCRIPTION
**Pull request by robertMihai19**
Add CF and OF exercise

### CODE REVIEW

File: laborator/content/introducere-asamblare/README.md [hunk 1]
```diff
@@ -256,6 +256,10 @@ Calculați minimul dintre numerele din 2 registre (eax și ebx) folosind o instr
 
 Calculați al N-lea număr Fibonacci, unde N este dat prin intermediul registrului `eax`.
 
+### 6. BONUS: Fibonacci
+
+Calculați al N-lea număr Fibonacci, unde N este dat prin intermediul registrului `eax`.
+
```
Comment (dbaluta) [inline laborator/content/introducere-asamblare/README.md:262]: There is something strange here. Commit subject says you add CF vs OF example, but here you also add another  Fibonacci exercise. 

This is confusing. A commit should only do one thing.

File: laborator/content/introducere-asamblare/7-cf-of/cf-of.asm [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+section .text
+    global main
+    extern printf
+
+main:
+    mov eax, 0x7FFFFFFF
+    jmp no_flag
+    
+no_flag:
+    PRINTF32 `CF si OF nu-s active\n\x0`
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf-of.asm:12]:
```suggestion
    PRINTF32 `CF si OF nu sunt active\n\x0`
```

File: laborator/content/introducere-asamblare/7-cf-of/cf-of.asm [hunk 3]
```diff
@@ -0,0 +1,33 @@
+%include "printf32.asm"
+
+section .text
+    global main
+    extern printf
+
+main:
+    mov eax, 0x7FFFFFFF
+    jmp no_flag
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf-of.asm:9]:
```suggestion
```
No need to `jmp` to a label just below the `jmp` instruction.

File: laborator/content/introducere-asamblare/README.md [hunk 4]
```diff
@@ -256,6 +256,10 @@ Calculați minimul dintre numerele din 2 registre (eax și ebx) folosind o instr
 
 Calculați al N-lea număr Fibonacci, unde N este dat prin intermediul registrului `eax`.
 
+### 7. BONUS: CF vs OF
+
+Folosind operatia de add pe registrul `eax`, setati OF, apoi CF si apoi ambele flaguri simultan.
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/README.md:261]:
```suggestion
Folosind instrucțiunea `add` pe registrul `al`, setați OF, apoi CF și apoi ambele flaguri simultan.
```

**Review by teodutu — CHANGES_REQUESTED** [review]
The code example is OK, but can be improved. See my inline suggestions, as well as the following:
- Don't make the setting of the 2 flags "cascade" from one another. Instead, create 3 independent snippets (in the same `.asm` file), one for each scenario. Use `test al, al` to clear both flags before each snippet
- The code you added looks more like the solution rather than the skeleton. The solution should be placed in `laborator/content/introducere-asamblare/solution/7-cf-of/`.
- Use 8-bit registers and values (`al` for example) so the numbers are easier to understand
- Mind the `checkpatch.pl` errors: remove the trailing whitespaces.

File: laborator/content/introducere-asamblare/7-cf-of/cf.asm [hunk 5]
```diff
@@ -0,0 +1,22 @@
+%include "printf32.asm"
+
+section .text
+    global main
+    extern printf
+
+main:
+    xor al, al
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf.asm:8]:
```suggestion
```
You reset `al` one line below.

File: laborator/content/introducere-asamblare/7-cf-of/cf.asm [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    mov al, 0xFF
+    
+no_flag1:
+    PRINTF32 `CF nu e activ\n\x0`
+    test al, al
+    ;TODO: activati CF
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf.asm:14]:
```suggestion
    ;TODO: activati CF

```

File: laborator/content/introducere-asamblare/7-cf-of/cf.asm [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+section .text
+    global main
+    extern printf
+
+main:
+    xor al, al
+    mov al, 0xFF
+    
+no_flag1:
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf.asm:11]:
```suggestion
```
You never `jmp` to this label.

File: laborator/content/introducere-asamblare/7-cf-of/cf.asm [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    PRINTF32 `CF nu e activ\n\x0`
+    test al, al
+    ;TODO: activati CF
+    jc carry_flag1
+    jmp end1
+
+carry_flag1:
+    PRINTF32 `CF activ\n\x0`
+
+end1:
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/cf.asm:21]:
```suggestion
    jc carry_flag
    jmp end

carry_flag:
    PRINTF32 `CF activ\n\x0`

end:
```

File: laborator/content/introducere-asamblare/solution/7-cf-of/Makefile [hunk 9]
```diff
@@ -0,0 +1,2 @@
+BINARIES := cf-of cf of
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/Makefile:1]: Turn this into a link to `../../7-cf-of`.

File: laborator/content/introducere-asamblare/7-cf-of/Makefile [hunk 10]
```diff
@@ -1,2 +1,2 @@
-PROGNAME := cf-of
-include ../../utils/Makefile.generic
+BINARIES := cf-of cf of
+include ../../utils/Makefile.single
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/7-cf-of/Makefile:2]:
```suggestion
include ../../utils/Makefile.multiple
```

**Review by teodutu — CHANGES_REQUESTED** [review]
It looks better now. Keep in mind to apply my inline comments from `cf.asm` to all code files.

**Comment by teodutu** [comment]
Also fix the `checkpatch.pl` errors. You have left some trailing whitespaces.

File: laborator/content/introducere-asamblare/solution/7-cf-of/Makefile [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+include ../../utils/Makefile.single
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/Makefile:2]:
```suggestion
include ../../utils/Makefile.multiple
```

File: laborator/content/introducere-asamblare/solution/7-cf-of/cf-of.asm [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+section .text
+    global main
+    extern printf
+
+main:
+   xor al, al
+   mov al, 0x1
+    
+no_flag:
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/cf-of.asm:11]:
```suggestion
```
Don't add labels that you don't use.

File: laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+section .text
+    global main
+    extern printf
+
+main:
+    xor al, al
+    mov al, 0xFF
+    
+no_flag1:
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm:11]:
```suggestion
```

File: laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    PRINTF32 `CF nu e activ\n\x0`
+    test al, al
+    add al, 1
+    jc carry_flag1
+    jmp end1
+
+carry_flag1:
+    PRINTF32 `CF activ\n\x0`
+
+end1:
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm:21]:
```suggestion
    jc carry_flag
    jmp end

carry_flag:
    PRINTF32 `CF activ\n\x0`

end:
```
No need to add a `1` at the end.

File: laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    extern printf
+
+main:
+    xor al, al
+    mov al, 0xFF
+    
+no_flag1:
+    PRINTF32 `CF nu e activ\n\x0`
+    test al, al
+    add al, 1
```
Comment (teodutu) [inline laborator/content/introducere-asamblare/solution/7-cf-of/cf.asm:14]: This is the same example as `cf-of.asm`. Use something else, such as `200 + 100` to highlight the difference.

**Review by teodutu — CHANGES_REQUESTED** [review]
It's better now. I have made some additional suggestions. In addition to them, you need to remove the binaries (`cf`, `of` and `cf-of`) from `7-cf-of/` and then add them to a `.gitignore` file in the same directory so you don't add them again.

And don't mind the `checkpatch.pl` errors about whitespaces. I think Assembly confused the script somehow :)))

**Review by teodutu — CHANGES_REQUESTED** [review]
The code looks OK. Just eliminate the trailing whitespace highlighted by `checkpatch.pl`. This GitHub action runs whenever you make a new commit and leaves inline messages [1, 2]. Check them out after you make changes. No need to wait for me to make the same suggestions.

[1] https://github.com/systems-cs-pub-ro/iocla/pull/139/files
[2] https://github.com/systems-cs-pub-ro/iocla/actions/runs/4430626762/jobs/7772605275?pr=139

**Review by teodutu — APPROVED** [review]
Now it's good.

**Pull request merged**